python/orca/tutorial/NCF/pytorch_dataset.py

```diff
@@ -22,70 +22,143 @@
+    items = pd.read_csv(
+        os.path.join(dataset_dir, 'movies.dat'),
+        sep="::", header=None, names=['item', 'category'],
+        usecols=[0, 2], dtype={0: np.int32, 1: str})
+
+    user_num = users['user'].max() + 1
+    item_num = items['item'].max() + 1
+
+    # categorical encoding
+    for i in sparse_features:
+        df = users if i in users.columns else items
+        df[i], _ = pd.Series(df[i]).factorize()
+    sparse_features.append('occupation')  # occupation is already indexed.
+
+    # scale dense features
+    for i in dense_features:
+        scaler = MinMaxScaler()
+        df = users if i in users.columns else items
+        values = df[i].values.reshape(-1, 1)
+        values = scaler.fit_transform(values)
+        values = [torch.tensor(v, dtype=torch.float32) for v in values]
+        df[i] = values
+
+    feature_cols = ['user', 'item'] + sparse_features + dense_features
+    label_cols = ["label"]
+    return users, items, user_num, item_num, \
+        sparse_features, dense_features, feature_cols+label_cols
+
+
+def get_input_dims(users, items, sparse_features, dense_features):
+    # Calculate input_dims for each sparse features
+    sparse_feats_input_dims = []
+    for i in sparse_features:
+        df = users if i in users.columns else items
+        sparse_feats_input_dims.append(df[i].max()+1)
+
+    num_dense_feats = len(dense_features)
+    return sparse_feats_input_dims, num_dense_feats
+
+
+def process_ratings(dataset_dir, user_num, item_num):
+    ratings = pd.read_csv(
         os.path.join(dataset_dir, 'ratings.dat'),
         sep="::", header=None, names=['user', 'item'],
-        usecols=[0, 1], dtype={0: np.int64, 1: np.int64})
-
-    user_num = data_X['user'].max() + 1
-    item_num = data_X['item'].max() + 1
+        usecols=[0, 1], dtype={0: np.int32, 1: np.int32})
 
     # load ratings as a dok matrix
-    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int64)
-    for x in data_X.values.tolist():
+    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int32)
+    for x in ratings.values.tolist():
         train_mat[x[0], x[1]] = 1
+    return ratings, train_mat
+
+
+def load_dataset(dataset_dir, num_ng=4):
+    """
+    dataset_dir: the path of the datasets;
+    num_ng: number of negative samples to be sampled here.
+    """
+    users, items, user_num, item_num, sparse_features, dense_features, \
+        total_cols = process_users_items(dataset_dir)
+    ratings, train_mat = process_ratings(dataset_dir, user_num, item_num)
+
+    # sample negative items
+    dataset = NCFData(ratings, item_num, train_mat, num_ng)
+    dataset.ng_sample()
+
+    # train test split
+    train_dataset, test_dataset = dataset.train_test_split()
 
-    return data_X, user_num, item_num, train_mat
+    # merge features
+    train_dataset.merge_features(users, items, total_cols)
+    test_dataset.merge_features(users, items, total_cols)
```

Comment:
first merge and then split?

Reply:
![image](https://user-images.githubusercontent.com/42887453/204937851-9269cdb4-70fa-4da6-97ef-774779300d21.png)
merge only after ratings is split.